### PR TITLE
implement tooltip follow cursor

### PR DESCRIPTION
**Pull request by senluchen2015**
- implement an option in tooltip that when followCursor: true, the tooltip will follow the cursor instead of the calculated position. The tooltip is positioned to the top-right of the cursor pointer. This should work with any chart and with any size tooltip.

Demo:
http://jsfiddle.net/3yqyvdud/2/

### CODE REVIEW

File: src/scripts/visualizations/tooltip.js [hunk 1]
```diff
@@ -47,6 +47,7 @@
             var pointX = xScale ? xScale(d.x) : pointOrCentroid.call(this)[0];
             var pointY = yScale ? yScale(d.y) : pointOrCentroid.call(this)[1];
             var alignedRight;
+            var followCursor = this.options.tooltip.followCursor;
```
Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: can you add an entry for `followCursor` under the 'defaults' in the beginning of this file? Listing it there is useful to quickly scan and get a comprehensive list of available options.

File: src/scripts/visualizations/tooltip.js [hunk 2]
```diff
@@ -97,10 +98,21 @@
                     };
 
                     return clampPosition(pos);
+                },
+
+                'cursor': function cursorPositioner() {
+                    var pos = d3.mouse(layer[0][0]);
+                    var tooltip = tooltipElement[0][0]
+                    var offsetX = 40;
```
Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: What's 40? Just an arbitrary number to distance from the cursor?

File: src/scripts/visualizations/tooltip.js [hunk 3]
```diff
@@ -97,10 +98,21 @@
...
+                    var offsetY = -(tooltip.clientHeight);
+                    return {
+                        x: pos[0] + offsetX,
+                        y: pos[1] + offsetY,
+                    };
                 }
             };
 
-            return options.chart.rotatedFrame ? positioner.horizontal() : positioner.vertical();
+            return options.tooltip.followCursor ? positioner.cursor() : (options.chart.rotatedFrame ? positioner.horizontal() : positioner.vertical());
```
Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: Can you refactor this so is this not a nested ternary?

**Comment by parikhshiv** [comment]
dude maybe I'm doing this wrong (in http://jsfiddle.net/3yqyvdud/2/), but when I change the distanceX / distanceY values, the tooltip's position doesn't move.

also, you can version your contour.js source code in jsfiddle by git commit - check out http://jsfiddle.net/q1psy5t3/2/ as an example

File: src/scripts/visualizations/tooltip.js [hunk 4]
```diff
@@ -112,6 +124,12 @@
             changeOpacity.call(this, 0, this.options.tooltip.hideTime);
         };
 
+        var onMouseMove = function (d) {
+            if (this.options.tooltip.followCursor) {
+                show.call(this, d)
```
Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: do you need the `call` here? (you might, just checking)

**Comment by parikhshiv** [comment]
FWIW, think this tool is super useful, and looks like it works great in the example. @narenranjit's comments make sense to me - I'll add my feedback to them directly.

File: src/scripts/visualizations/tooltip.js (hunk 2 above)
Comment (parikhshiv) [inline src/scripts/visualizations/tooltip.js]: I see why you need to pick some distance though. Maybe you could just mark this as a constant (all caps, snake case is preferred notation in pre-es6 javascript **I think).

Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: My question was more about does this magic number work across all font-sizes/other stuff we don't know about or should we make this configurable

File: src/scripts/visualizations/tooltip.js (hunk 4 above)
Comment (senluchen2015) [inline src/scripts/visualizations/tooltip.js]: I think so, need the "this" in the show function. Alternatively can do ```show.bind(this)(d)``` but I think call is better.

Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: It looks like it'll have the right `this` even without call anyway - unless d3/someone else is hijacking it. Minor comment though, ignore if it doesn't work without call.

File: src/scripts/visualizations/tooltip.js (hunk 2 above)
Comment (senluchen2015) [inline src/scripts/visualizations/tooltip.js]: It's the magic offset number that position the tooltip nicely on the in the horizontal position regardless of the size of the tooltip.

**Comment by narenranjit** [comment]
+1 to adding a rawgit link (i.e., update dist/contour as part of this PR and use rawgit to link to that commit) to jsfiddle

File: src/scripts/visualizations/tooltip.js (hunk 2 above)
Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: sounds fine then

**Comment by senluchen2015** [comment]
Updated jsfiddle with rawgit: http://jsfiddle.net/3yqyvdud/3/

File: src/scripts/visualizations/tooltip.js [hunk 5]
```diff
@@ -9,7 +9,8 @@
             distance: 5,
             distanceX: undefined,
             distanceY: undefined,
```
Comment (narenranjit) [inline src/scripts/visualizations/tooltip.js]: Would it make sense to say distance can be a number or an object of the form {x :<x>, y: <y>}? That way we'll never have overlapping contradictory values

File: src/scripts/visualizations/tooltip.js (hunk 2 above)
Comment (drosen0) [inline src/scripts/visualizations/tooltip.js]: Looking at this code, I think the same thing: what is this magic number? The only way a developer could find the answer is by finding and reading this PR. It'd be better to make it self-documenting, and might as well leave a comment at the same time:

```js
// magic offset number that position the tooltip nicely on the
// horizontal position regardless of the size of the tooltip
var HORIZONTAL_TOOLTIP_OFFSET = 40;
var offsetX = HORIZONTAL_TOOLTIP_OFFSET;
```

**Comment by senluchen2015** [comment]
Merged with tooltip distance and made followCursor default to true.

Demo: http://jsfiddle.net/3yqyvdud/4/

**Comment by narenranjit** [comment]
@senluchen2015 This looks okay, but feels sluggish (especially the bar) in the jsfiddle, and I'm not sure why.. is it worth trying to delegated event handlers to see if that makes a difference? Or is that too much work?

**Pull request closed without merging**